### DISCUSSION TITLE
Build problem after update to current version and update of tools environment

### DISCUSSION THREAD

**Discussion by mg4gh** [mapsforge/mapsforge#1291] [Q&A] [answered]
I'm using mapsforge for my android app, but I was  using still version 0.12 upto now.
Yet I tried to upgrade and then I needed also a new Version of AndroidStudio (now 2021.1.1 build from 1.2.2022) and Gradle Version 7.2.

Now I get a mess of compile errors: 
error: package android.annotation does not exist
error: package android.os does not exist
error: package android.graphics.Bitmap does not exist
...
So in fact the compiler doesn't find the basic android packages. Any hint what could be wrong with my installation???
(I tried to uninstall/install the AndroidStudio and also the sdk packages, but no change).

Thanks in advance, Martin

PS: Since I build some jars from mapsforge I still can build my app with these jars also with the new Studio and Gradle version.

**Comment by devemux86** [comment, marked as answer]
Set an environment variable `ANDROID_HOME` with the path to the Android SDK installation directory.
https://developer.android.com/studio/command-line/variables

---

Mapsforge releases and snapshots are available in Maven and Sonatype,
see the [integration guide](https://github.com/mapsforge/mapsforge/blob/master/docs/Integration.md).

**Comment by mg4gh** [comment]
Thanks, but setting ANDROID_HOME doesn't seem to effect the behavior. Anyway the SDK Location as reflected in the Project Structure seems to be ok. I would prefer not to use prebuild jars, since (at least in the past) I wanted to change a few lines (as done also with the contributions to your project). 
The problem seems somehow related to my fork of the mapsforge project, since e.g. the compiler claims 
> AndroidBitmap.java:23: error: package android.os does not exist import android.os.Build;
(and it even doesn't find  java classes from java.io).
On the same version of AndroidStudio I can compile my App which is using the same method to check the Build.VERSION, so why does the build system in the mapsforge project claims an error while the same code in another project doesn't?
Further, when I clone both projects from github fresh, then I run into the same problem with the mapsforge project, my project is running well. 
Gradle is using the java11 as delivered with AndroidStudio, but there is no difference bewteen both projects.

Any further idea what's going wrong with my mapsforge clone?

**Comment by mg4gh** [comment]
Sorry, I was wrong - I set ANDROID_HOME, but not the way that AndroidStudio recognized it. I saw this with a debug statement in the gradle file, so now it compiles :-) Nevertheless I wonder whether this dependency is necessary.

**Comment by mg4gh** [comment]
There is still one issue: Although I can build and run the sample app, if I go into the sources, I get plenty of error - typically with basic java classes, but also with those references to other classes that reference these classes. E.g. AS failes with AndroidGraphicFactory because GraphicFactory fails to be compiled, because this AS doesn't find the "import java.io.IOException" and "import java.io.InputStream".

Adding the line:
    compileOnly files("${System.env.ANDROID_HOME}/platforms/android-${androidCompileSdk()}/android.jar")
to the shared modules (shard between AWT and android) solves this issue. But I guess that you added this line only to pure android modules by intention, so does this line in shared modules violate the multi platform approach?

**Reply by devemux86** [reply]
Android modules use the Android jar.
The shared modules work with Java.

Set "ANDROID_HOME" variable, clone the repository, open with IDEA or build with Gradle, everything works.
Perhaps a "JAVA_HOME" variable is also needed in your environment.

**Reply by mg4gh** [reply]
Unfortunately it's not working this way. I tried also to fresh clone from github to make sure that I have not an issue with some cache. But I could reproduce the problem. ANDROID_HOME is set and the total build of the mapsforge-samples-android is working well since this env variable is set. Also I can run the sample app in the emulator.   
The JAVA_HOME was originally set to a jdk8, then I tried to set it to the java11 as it is dristributed with the AndroidStudio and used by gradle. 
Still the problem remains: the shared modules do not find the basic java classes.
<img width="1305" alt="as" src="https://user-images.githubusercontent.com/30026615/153606981-1f846536-4b16-458c-8f38-36cd45888cc8.PNG">

**Reply by devemux86** [reply]
Android Studio's Java is for Android Studio.

You need Java installed and properly configured in your operating system for IDE to use it in the projects.

**Reply by mg4gh** [reply]
As I said, initially I tried with an jdk8 - no success. The with the version from AS - still no success. 
Now I changed again to a jdk13 that is also installed - again no success.
If I go to the terminal tab in AS: then "java -version" works fine
<img width="1334" alt="as2" src="https://user-images.githubusercontent.com/30026615/153612020-825fd9d3-1d89-40e5-b18c-744b5a55297a.PNG">
.

**Reply by devemux86** [reply]
I cloned the repository and opened the project in Android Studio and IDEA with clean settings.
Everything works correctly.

Probably something is needed to be fixed in your Java or IDE.

**Reply by mg4gh** [reply]
What version of AndroidStudio did you use?
I've tried a very new one:
<img width="482" alt="as3" src="https://user-images.githubusercontent.com/30026615/153648589-d2340c1d-cd08-4670-8aa6-3c38260a0d01.PNG">

**Reply by mg4gh** [reply]
I've also and old IDEA installation, so after two updates I'm at version 2021.3.2 - and this is woking too.
So it looks like the AndroidStudio (with above version) is causing the problem.
<img width="1264" alt="as4" src="https://user-images.githubusercontent.com/30026615/153651550-778db02d-9d8b-464a-8a8d-c95b946df2dc.PNG">
.

**Reply by devemux86** [reply]
> What version of AndroidStudio did you use?

Latest stable releases:
- Android Studio Bumblebee 2021.1.1 Patch 1
- IntelliJ IDEA 2021.3.2

Android Studio is not required for Mapsforge, it works with IDEA too.

If you use the IDE via snap,
reinstall the IDE from tar package downloaded from IDE website, works better.

**Reply by mg4gh** [reply]
So it's exactly the same version of AndroidStudio. The ANDRIOS_HOME is correct, otherwise the sample app would not run. THe JAVA_HOME seems to be ok, otherwise IDEA should not run. 
Nevertheless, even with a fresh clone from github I run into the same problem.
But since I have with IDEA a way that works, it dosn't matter much. Thanks again for investigations.

**Comment by devemux86** [comment]
If you check the GitHub Actions, the repository is built successfully:
https://github.com/mapsforge/mapsforge/actions
